modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/twostep/AbstractReducer.java

```diff
@@ -319,20 +320,23 @@ protected final ReduceResultPage createDummyLastPage(ReduceResultPage lastPage)
      * @param iter Current iterator.
      * @return The same or new iterator.
      */
-    protected final Iterator<Value[]> pollNextIterator(Pollable<ReduceResultPage> queue, Iterator<Value[]> iter) {
-        if (!iter.hasNext()) {
+    protected final IgniteIntObjectTuple<Iterator<Value[]>> pollNextIterator(
+        Pollable<ReduceResultPage> queue,
+        IgniteIntObjectTuple<Iterator<Value[]>> iter
+    ) {
+        if (!iter.get2().hasNext()) {
             try (TraceSurroundings ignored = MTC.support(ctx.tracing().create(SQL_PAGE_FETCH, MTC.span()))) {
                 ReduceResultPage page = takeNextPage(queue);
 
                 if (!page.isLast())
                     page.fetchNextPage(); // Failed will throw an exception here.
 
-                iter = page.rows();
+                iter = F.intt(page.columnCount(), page.rows());
```

Comment:
`page.columnCount()` is the same for all pages, looks like it's more correct to store it in reducer. Also it will reduce GC pressure. WDYT?

Reply:
In the case of distributed query - empty page can be received from some node (columnCount == 0) and the other page with requested column count will be recieved afterwards

Reply:
This field is accessed only when `iter.hsNext()` returns `true`. You can store first non-zero value.